src/realm.js

```diff
@@ -1485,6 +1490,13 @@ export class Realm {
         });
       }
     }
+
+    // materialize leaked objects
+    if (leakedObjects !== undefined && leakedObjects.size !== 0) {
+      for (let lo of leakedObjects) {
+        Materialize.materialize(this, lo);
```

Comment:
Is this the main place where materialization happens when joining states?

This code looks dangerous to me. It would permanently mutate whatever generator was captured in `lo`, right? Shouldn't rather mutate the current `generator`? Or is there an invariant that all those generators are identical?

Reply:
@hermanventer ?

Reply:
I'm trying to move towards a world where Effects is immutable. This code certainly will not help me get there. I think it should be OK to materialize the leaked objects in the the current generator at this point. Of course, only do so if the appendGenerator flag is true.

Hopefully this will allow the code to become simpler, which is something I'm always in favor of.

Reply:
I considered several alternatives, and the motivation for capturing the generator at the leak point, creating a LeakedObject type, and modifying the Effects structure was to pass it here, as this is the point at which all of the information needed for materialization comes together. If you materialize into the current generator at this point, you will see the object being built post-join. If you materialize where bindings and properties are joint, you will see disembodied objects missing property bindings, and unexpected deletes because descriptors that `applyEffects` eventually populates are not present at that point. This deletion logic is quite involved and I am glad I avoided colliding with it with the approach that I took.

Reply:
I have added a comment to that code.